Block conversion update: ability to convert objects with advanced configuration

VolgaIgor opened the pull request:
Hello!
Right now block conversions are performed using a string only, which does not allow for complex conversions.
For example, it will not be possible to convert text blocks separately and image blocks separately, like this:
![image](https://github.com/user-attachments/assets/7eb8c304-72b2-48e8-87b8-bc45bcf1e99e)

Right now it is not possible to send a complex configuration from an image block to a gallery block (file data + caption) via string. In addition, there is no conversion filtering now, so if a block exports something, it will be listed for import in all blocks, even if it is not suitable for conversion.

I made a few changes to make it all possible:
* The conversion menu is not displayed for blocks that do not have the "export" option (which do not export anything): https://github.com/codex-team/editor.js/pull/2799
* A block can now return for export not only a string, but also a data object
* The conversion result is now computed when the conversion menu is opened. so if a block returns an empty object or undefined when importing from another block, it will not be suggested for conversion
* Backwards compatibility with past mechanics has been preserved as much as possible

VolgaIgor committed:
Fixed display of convert menu for blocks without export rule

According to the workflow script from the documentation:
https://editorjs.io/tools-api/#conversionconfig

VolgaIgor committed:
Block conversion update: ability to convert to complex objects

VolgaIgor committed:
Block conversion update: ability to convert to complex objects

neSpecc commented:
Thanks for the PR. Could you show the example of the conversion configs in Image Tool and Gallery tools?

VolgaIgor commented:
> Thanks for the PR. Could you show the example of the conversion configs in Image Tool and Gallery tools?

For example, the SimpleImage export I did it this way:
```js
export: function (blockData) {
  return {
    'url': blockData.url,
    'caption': blockData.caption
  }
}
```

And for ImageTool this way:
```js
export: function (blockData) {
  return {
    'file': blockData.file,
    'caption': blockData.caption
  }
}
```

And for Gallery I made a universal way to handle both variants, otherwise returning undefined:
```js
import: function (exportData) {
  if (typeof exportData !== 'object') return;

  let blockData = {};
  if (exportData.url) {
    blockData.files = [{ url: exportData.url }];
  } else if (exportData.file) {
    blockData.files = [exportData.file];
  } else {
    return;
  }

  if (exportData.caption) {
    blockData.caption = exportData.caption;
  }
  return blockData;
}
```

neSpecc commented:
The problem now is that Gallery tool in your example knows about other tools: SimpleImage and ImageTool. It is a bad practice for our design. That is why we designed conversion through the string data.

As a solution, we can improve ImageTool by supporting multiple images in a gallery-view.

VolgaIgor commented:
I interpret this not as the block knows about these tools, but that this block is ready to accept data in a certain format. And Image blocks give their data in a more structured format.

So it doesn't mean that Gallery only works with these Image blocks, and that Image blocks will only work with Gallery. They will only be able to have more extended interfaces for sharing structured data.

VolgaIgor commented:
Or maybe do something similar to Clipboard, allowing export to text/plain and text/html. HTML5 is advanced enough that it can be used to explicitly deliver various data structures, including lists, image galleries, etc.
@neSpecc

VolgaIgor committed:
Merge branch 'next' into new-conversion

VolgaIgor committed:
Merge fix

neSpecc commented:
> Or maybe do something similar to Clipboard, allowing export to text/plain and text/html. HTML5 is advanced enough that it can be used to explicitly deliver various data structures, including lists, image galleries, etc. @neSpecc

I don't see any options at the moment. Transferring data should be abstracted away from any tool's data format. Maybe markdown would work better that just simple strings. We'll think about it when we'll design a markdown support (in Tools API first of all)

neSpecc committed:
fix(inline-tools): inline tools shortcuts now works in read-only mode (#2891)

* fix(inline-tools): inline tools shortcuts now works in read-only mode

* use ubuntu-20.04 instead of latest

neSpecc committed:
chore(submodules): update all submodules to the latest version (#2880)

* update all submodules

* rm checklist and nested list submodule

* Update .gitmodules

* rm list submodule

* add list submodule

* all submodules updated

* fix(inline-tools): inline tools shortcuts now works in read-only mode

github-actions[bot] committed:
Bump version (#2895)

Co-authored-by: github-actions <action@github.com>

idebenone committed:
fix: properly handle visible whitespaces in empty blocks (#2865)

* fix: handle whitespace input in empty placeholders correctly

* fix: isNodeEmpty() to handle visible whitespaces

* chore: bump version from 2.31.0-rc.5 to 2.31.0-rc.6

* chore: bump version from 2.31.0-rc.5 to 2.31.0-rc.6

* fix: submodules updated

* fix: eslint errors

* test: backspace removes trailing spaces, hides placeholder in empty blocks

* fix: update incorrect tests

* fix: resolving submodules issue

* Create list

---------

Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

idebenone committed:
chore: bump version for release (#2896)

neSpecc committed:
dx(submodules): remove tools submodules from the repo (#2898)

* rm all submodules

* rm commands

* Update package.json

obafemitayor committed:
Bug Fix For When "/" Overides external text (#2894)

* Bug Fix For When / Overides external text

* Moved fix to blockEvents

* Moved fix to blockEvents

* Moved fix to blockEvents

* Refactored test to simulate behaviour

* Added fix to change log

* Refactored test to mimick exact behaviour of the bug

---------

Co-authored-by: Omotayo Obafemi <omotayo@testlio.com>
Co-authored-by: Peter <specc.dev@gmail.com>

obafemitayor committed:
Added fix for memory leak issue (#2893)

* Added fix for memory leak issue

* Documented the fix in docs/CHANGELOG.md

* v2.31.0

* Documented the fix in docs/CHANGELOG.md

* Documented the fix in docs/CHANGELOG.md

* Documented the fix in docs/CHANGELOG.md

* Documented the fix in docs/CHANGELOG.md

---------

Co-authored-by: Omotayo Obafemi <omotayo@testlio.com>
Co-authored-by: Peter Savchenko <specc.dev@gmail.com>

github-actions[bot] committed:
Bump version (#2899)

Co-authored-by: github-actions <action@github.com>

neSpecc committed:
imrovement(flipper): allow to select next/prev line by shift arrows (#2918)

* fix: prevent flipper navigation when shift key is pressed

* rm logs

* feat: improve line selection with Shift + Up/Down

* fix lint action

* fix action

* upd

github-actions[bot] committed:
Bump version (#2919)

Co-authored-by: github-actions <action@github.com>

neSpecc committed:
improvement(caret): caret.setToBlock() offset argument improved (#2922)

* chore(caret): caret.setToBlock offset improved

* handle empty block

* Update caret.cy.ts

* fix eslint

neSpecc committed:
Create .nvmrc

narpat-ps committed:
resolve "Can't find a Block to remove" error in renderFromHTML (#2941)

* fix(blocks):Error occurred when calling renderFromHTML: Can't find a Block to remove.

* fix: resolve "Can't find a Block to remove" error in renderFromHTML

- Make renderFromHTML async and await BlockManager.clear() to prevent race condition
- Change removeBlock order: remove from array before destroy to prevent index invalidation
- Fix clear() method to copy blocks array before iteration to avoid modification during loop

Fixes issue where renderFromHTML would fail with "Can't find a Block to remove" error
due to concurrent block removal operations and array modification during iteration.

Resolves #2518

neSpecc committed:
add changelog

neSpecc committed:
release 2.31 (#2956)

neSpecc committed:
add example of how to use i18n for Convert To buttons

VolgaIgor committed:
Merge branch 'codex-team:next' into new-conversion

VolgaIgor committed:
Fixed merge mistake
